File: src/hivemind/state.py
from copy import deepcopy
import logging
from typing import Iterator, List, Set, Tuple
from functools import cached_property
import random
import json

from .insect import Insect, Team, Stone
from .hive import Hive
from .hex import Hex

logger = logging.getLogger(__name__)


class Action:
    """ Base class for abstract game action that can be performed in a turn """
    def __eq__(self, other: "Action") -> bool:
        return self.__dict__ == other.__dict__

    def __repr__(self) -> str:
        args = ", ".join((v.__repr__() for v in self.__dict__.values()))
        return f"{type(self).__name__}({args})"

class Move(Action):
    """ Move from a origin Hex to a destination Hex """
    def __init__(self, origin: Hex, destination: Hex):
        self.origin = origin
        self.destination = destination


class Drop(Action):
    """ Drop a stone to a destination Hex """
    def __init__(self, stone: Stone, destination: Hex):
        self.stone = stone
        self.destination = destination

class Pass(Action):
    """ When no action is possible you have to pass """
    pass


class State:
    """
    State contains all information to completely describe each state of the game
    and some helper functions/attributes to optimize the processing.
    Necessary information is:
    - hive: Dict[Hex: List[Stone]]
    - turn_number: int
    Derivable from the atomic information
    - _bee_move
    - availables

    Only valid states are representable.
    Root is the only starting point and every state is an ancestor of it.
    Any child is yielded from the application of a valid action on the parent state.
    """

    def __repr__(self):
        return f"State({self.hive}, {self.turn_number})"

    def to_json(self):
        # TODO
        dump = {}
        dump["hive"] = []
        for hex, stack in self.hive.items():
            for height, stone in enumerate(stack):
                dump["hive"].append({})
                temp = dump["hive"][-1]
                # r, s, h, name, team
                temp["q"] = hex.q
                temp["r"] = hex.r
                temp["height"] = height
                temp["name"] = stone.insect.value
                temp["team"] = stone.team.value
        dump["availables"] = []
        for stone in self.availables:
            dump["availables"].append({})
            temp = dump["availables"][-1]
            temp["name"] = stone.insect.value
            temp["team"] = stone.team.value

        # TODO: add more information
        return json.dumps(dump)

    @property
    def current_team(self) -> Team:
        return Team.WHITE if self.turn_number % 2 else Team.BLACK

    @property
    def bee_move(self) -> bool:
        return self._bee_move[self.current_team.value]

    def __add__(self, action: Action) -> "State":
        """ Returns a new State with the action performed """
        assert action in self.possible_actions
        new_state = deepcopy(self)
        new_hive = new_state.hive
        if isinstance(action, Move):
            # Remove the stone from the old position and add it at the new one
            stone = new_hive.at(action.origin)
            new_hive.remove_stone(action.origin)
            new_hive.add_stone(action.destination, stone)
        elif isinstance(action, Drop):
            stone = action.stone
            if stone.insect == Insect.BEE:
                # Update that the bee is dropped
                new_state._bee_move[self.current_team.value] = True
            # Remove the dropped stone from the availables and add it to the hive
            new_state.availables.remove(stone)
            new_hive.add_stone(action.destination, stone)
        new_state.turn_number += 1
        # Unset them so they are recomputed on the new state
        if hasattr(new_state, "possible_actions"):
            del new_state.__dict__["possible_actions"]
        logger.debug(f"Created new state {new_state}")
        return new_state

    @property
    def game_result(self):
        return self.hive.game_result

    def is_game_over(self) -> bool:
        """ Check if the game is over """
        return not (self.game_result is None)

    def _unique_availables(self) -> Set[Stone]:
        """ Returns the unique availables Stones that can be dropped by the current team """
        return {a for a in self.availables if a.team == self.current_team}

    @cached_property
    def possible_actions(self) -> Tuple[Action]:
        """ Generate all legal actions for the current state """
        opts = []
        drop_stones = self._unique_availables()
        if self.turn_number == 0:
            opts = [Drop(stone, Hex(0, 0)) for stone in drop_stones]
        elif self.turn_number == 1:
            neighbours = self.hive._get_root().neighbours()
            opts = [Drop(stone, hex) for stone in drop_stones for hex in neighbours]
        elif self.turn_number >= 6 and not self.bee_move:
            for drop_hex in self.hive.generate_drops(self.current_team):
                opts.append(Drop(Stone(Insect.BEE, self.current_team), drop_hex))
        else:
            for drop_hex in self.hive.generate_drops(self.current_team):
                opts.extend(Drop(stone, drop_hex) for stone in drop_stones)
            if self.bee_move:
                for origin, destination in self.hive.generate_moves(self.current_team):
                    opts.append(Move(origin, destination))
        return tuple(opts) if opts else (Pass(),)

    def children(self) -> Tuple[Action]:
        """ Returns a tuple of all possible child states """
        return tuple(self + action for action in self.possible_actions)

    def next_state(self, policy=random.choice) -> "State":
        """ Takes a policy function that has to select an action out of the tuple ot possibles """
        return self + policy(self.possible_actions)


class Root(State):

    def __init__(self):
        self.hive = Hive()
        self._bee_move = [False, False]
        self.turn_number = 0
        insects = (Insect.BEE,
                   Insect.SPIDER, Insect.SPIDER,
                   Insect.ANT, Insect.ANT, Insect.ANT,
                   Insect.GRASSHOPPER, Insect.GRASSHOPPER, Insect.GRASSHOPPER,
                   Insect.BEETLE, Insect.BEETLE)
        self.availables = [Stone(insect, team) for insect in insects for team in list(Team)]

File: src/shared/model/state.js
import {insects} from "./insects.js"
import {Stone} from "./stone.js"
import {teams} from "./teams.js"
import * as HEX from "../hexlib.js"
import {Move, Drop, Pass} from "./action.js"
import {Hive} from "./hive.js"

const startingInsects = [
    insects.BEE,
    insects.SPIDER,
    insects.SPIDER,
    insects.ANT,
    insects.ANT,
    insects.ANT,
    insects.GRASSHOPPER,
    insects.GRASSHOPPER,
    insects.GRASSHOPPER,
    insects.BEETLE,
    insects.BEETLE
]

export class State {
    // was constructed with hive as argument once
    constructor(turnNumber = 0) {
        this.hive = new Hive();
        this.turnNumber = turnNumber
        this.stones = [];
        this._beeMove = new Map([
            ["WHITE", false],
            ["BLACK", false]
    ])
        for (const team in teams) for (const insect of startingInsects) {
            this.stones.push(new Stone(insect, team))
        }
    }

    get team() {
        return this.turnNumber % 2 === 0 ? teams.WHITE : teams.BLACK
    }
    get result() {
        return this.hive.gameResult()
    }
    get moveAllowed() {
        return this._beeMove.get(this.team)
    }
    generateDrops() {
        if (this.turnNumber === 0) return [HEX.Hex(0, 0)];
        else if (this.turnNumber === 1) return [HEX.Hex(0, -1)];
        return this.hive.generateDrops(this.team)
    }
    _getActions() {
        console.log("Generating actions")
        let opts = []
        const dropStones = this.stones.filter(stone => stone.team === this.team)
            .map(i => JSON.stringify(i))
            .filter((s, i, r) => r.indexOf(s) === i)
            .map(i => JSON.parse(i))
            .map(({insect, team}) => new Stone(insect, team))
        if (dropStones.length) {
            console.log("Drops allowed")
            this.generateDrops().forEach(d => dropStones.forEach(ds => opts.push(new Drop(ds, d))))
        }
        else if (this.turnNumber >= 6 && !this.moveAllowed) {
            console.log("Forced bee drop")
            this.hive.generateDrops(this.team).forEach(d => opts.push(new Drop(new Stone(insects.BEE, this.team), d)))
        } else {
            if (this.moveAllowed) {
                console.log("Moves allowed")
                this.hive.generateMoves(this.team).forEach(([origin, dest]) => {
                    opts.push(new Move(origin, dest))
                })
            }
        }
        if (opts.length) return opts
        else return [Pass]
    }
    get actions() {
        if (!this._actions) this._actions = this._getActions()
        return this._actions
    }

    isLegal(action) {
        return action in self.actions
    }
    apply(action) {
        // No check for legal action!
        // TODO apparently copying/cloning objects is not trivial in JS
        // So maybe just create new instance with the data as arguments
        // let newState = Object.assign(Object.create(Object.getPrototypeOf(this)), this)
        // State is mutable now, apply updates inplace instead of returning a new instance
        console.log(`Applying ${JSON.stringify(action)}`)
        let stone
        if (action instanceof Move) {
            // Remove the stone from the old position and add it at the new one
            stone = this.hive.at(action.origin)
            this.hive.removeStone(action.origin)
            this.hive.addStone(action.destination, stone)
        } else if (action instanceof Drop) {
            stone = action.stone
            if (stone.insect === insects.BEE) {
                // Update that the bee is dropped
                this._beeMove.set(this.team, true)
            }
            // Remove the dropped stone from the availables and add it to the hive
            // TODO the stone is not removed because objects do not compare equal for values
            let index = -1;
            this.stones.forEach((s, i) => {
              if (JSON.stringify(s) === JSON.stringify(stone)) index = i
            })
            this.stones.splice(index, 1)
            this.hive.addStone(action.destination, stone)
        }
        delete this._actions
        this.turnNumber++
    }

    step(policy=randomPolicy) {
        return this.apply(policy(this.actions))
    }
}

function randomPolicy(actions) {
    return actions[Math.floor(Math.random() * actions.length)]
}